CARDS-1303 - Add the ability to display informative text in questionnaires

marta- opened the pull request:
To test, use the questionnaire wizard to add "Information" boxes to questionnaires, then instantiate the questionnaire to see how it looks to the user filling the form. Try both non-paginated and paginated questionnaires (e.g, where one page contains only an Information card).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/FormEntry.jsx:
```diff
@@ -159,5 +176,7 @@ let displaySection = (sectionDefinition, path, depth, existingAnswer, key, onCha
     return displayQuestion(entryDefinition, path, existingAnswers, keyProp, classes, onAddedAnswerPath, sectionAnswersState, onChange, pageActive, isEdit, instanceId);
   } else if (SECTION_TYPES.includes(entryDefinition["jcr:primaryType"])) {
     return displaySection(entryDefinition, path, depth, existingAnswers, keyProp, onChange, visibleCallback, pageActive, isEdit, instanceId);
+  } else if (INFO_TYPES.includes(entryDefinition["jcr:primaryType"])) {
+    return displayInformation(entryDefinition, keyProp, isEdit);
```
**Note**: This code should be replaced with a component manager architecture (in the next iteration).

marta- replied:
Refactoring FormEntry to be less spaghetti is a bit more tedious and out of scope.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -183,12 +185,14 @@ let QuestionnaireItemSet = (props) => {
-        Object.entries(data).filter(([key, value]) => (value['jcr:primaryType'] == 'cards:Section' || value['jcr:primaryType'] == 'cards:Question'))
+        Object.entries(data).filter(([key, value]) => ENTRY_TYPES.includes(value['jcr:primaryType']))
             .map(([key, value]) =>
                 value['jcr:primaryType'] == 'cards:Question' ?
                 <Grid item key={key}><Question data={value} onActionDone={onActionDone} classes={classes}/></Grid> :
                 value['jcr:primaryType'] == 'cards:Section' ?
-                <Grid item key={key}><Section data={value} onActionDone={onActionDone} classes={classes}/></Grid>
+                <Grid item key={key}><Section data={value} onActionDone={onActionDone} classes={classes}/></Grid> :
+                value['jcr:primaryType'] == 'cards:Information' ?
+                <Grid item key={key}><Information data={value} onActionDone={onActionDone} classes={classes}/></Grid>
                 : null
```
**Note**: This code should be replaced with a component manager architecture (in the next iteration).

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/CreationMenu.jsx:
```diff
@@ -67,6 +67,7 @@ let CreationMenu = (props) => {
       >
           <MenuItem onClick={() => { openDialog('Question'); handleCloseMenu(); }}>Question</MenuItem>
           <MenuItem onClick={() => { openDialog('Section'); handleCloseMenu(); }}>Section</MenuItem>
+          <MenuItem onClick={() => { openDialog('Information'); handleCloseMenu(); }}>Information</MenuItem>
```
**Note**: This should iterate through ENTRY_TYPES rather than hardcoding them.

marta- committed:
CARDS-1303 - Add the ability to display informative text in questionnaires

sdumitriu reviewed (commented):
Looks good on first pass. I think you should add the new node type as a possible child type for section and questionnaire in the .cnd.

marta- committed:
CARDS-1303: Add the ability to display informative text in questionnaires
Refactoring to reduce spaghetti code

marta- committed:
CARDS-1303: Add the ability to display informative text in questionnaires
Addressing review comments: specified `cards:Information` as possible child of `cards:Section`

sashaandjic commented:
Looks and feels good as long as the description is on one or max two lines. Larger descriptions do not look good in the selection dialog nor in the paginated form as it takes to much room.
![image](https://user-images.githubusercontent.com/23243244/128562992-f513db45-3230-46d0-b696-5bb9d09dd0ec.png)
![image](https://user-images.githubusercontent.com/23243244/128563016-a67584bf-bf2e-44e4-af22-d28bfcdfad0d.png

marta- committed:
CARDS-1303: Add the ability to display informative text in questionnaires
Addressing issue identified by @sashaandjic : Information cards that are direct children of a paginated questionnaire show up on all pages

sdumitriu reviewed (commented):
Are info elements inside sections not displayed on purpose? E.g. add a new Information element inside Patient Information / Surveillance.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
info is not shown when placed inside the section.

sashaandjic commented:
> Are info elements inside sections not displayed on purpose? E.g. add a new Information element inside Patient Information / Surveillance.

 I think this is a bug.

marta- commented:
> Are info elements inside sections not displayed on purpose? E.g. add a new Information element inside Patient Information / Surveillance.

No, forgot to push one of the changes. I amended the commit.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -183,14 +189,11 @@ let QuestionnaireItemSet = (props) => {
+        Object.entries(data).filter(([key, value]) => ENTRY_TYPES.includes(value['jcr:primaryType']))
+            .map(([key, value]) => (
+                    EntryType => <Grid item key={key}>{React.createElement(eval(EntryType),  {data: value, onActionDone: onActionDone, classes: classes})}</Grid>
+                  )(_stripCardsNamespace(value['jcr:primaryType']))
```
This also works:

```suggestion
                    EntryType => <Grid item key={key}><EntryType data={value} onActionDone={onActionDone} classes={classes} /></Grid>
                  )(eval(_stripCardsNamespace(value['jcr:primaryType'])))
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/Questionnaire.jsx:
```diff
@@ -203,6 +206,42 @@ QuestionnaireItemSet.propTypes = {
+};
+
+Information.propTypes = {
+  onActionDone: PropTypes.func,
```
Unrelated, but the `propTypes` for `Question` uses the deprecated `closeData` name for this, and for `Section` it does not mention it at all. Should be updated at some point.

sdumitriu committed:
[misc] Replace explicit use of React.createElement with JSX syntex

marta- committed:
CARDS-1303: Add the ability to display informative text in questionnaires

Specified `cards:Information` as possible child of `cards:Questionnaire`

sashaandjic reviewed (approved):
tested - looking good
